src/libaktualizr/package_manager/packagemanagerfactory.cc

```diff
@@ -1,55 +1,38 @@
 #include "package_manager/packagemanagerfactory.h"
-#include "package_manager/packagemanagerfake.h"
 
-#ifdef BUILD_OSTREE
-#include "package_manager/ostreemanager.h"
-#ifdef BUILD_DOCKERAPP
-#include "package_manager/dockerappmanager.h"
-#endif
-#endif
+#include "logging/logging.h"
 
-#ifdef BUILD_DEB
-#include "package_manager/debianmanager.h"
-#endif
+static std::list<std::pair<std::string, PackageManagerBuilder>> *registered_pkgms_;
 
-#if defined(ANDROID)
-#include "package_manager/androidmanager.h"
-#endif
+bool PackageManagerFactory::registerPackageManager(const char *name, PackageManagerBuilder builder) {
+  // this trick is needed so that this object is constructed before any element
+  // is added to it
+  static std::list<std::pair<std::string, PackageManagerBuilder>> rpkgms;
 
-#include "logging/logging.h"
+  if (!registered_pkgms_) {
+    registered_pkgms_ = &rpkgms;
+  }
+  registered_pkgms_->emplace_back(name, builder);
+  return true;
+}
 
 std::shared_ptr<PackageManagerInterface> PackageManagerFactory::makePackageManager(
-    const PackageConfig& pconfig, const BootloaderConfig& bconfig, const std::shared_ptr<INvStorage>& storage,
-    const std::shared_ptr<HttpInterface>& http) {
-  switch (pconfig.type) {
-    case PackageManager::kOstree:
-#ifdef BUILD_OSTREE
-      return std::make_shared<OstreeManager>(pconfig, bconfig, storage, http);
-#else
-      throw std::runtime_error("aktualizr was compiled without OStree support!");
-#endif
-    case PackageManager::kDebian:
-#ifdef BUILD_DEB
-      return std::make_shared<DebianManager>(pconfig, bconfig, storage, http);
-#else
-      throw std::runtime_error("aktualizr was compiled without debian packages support!");
-#endif
-    case PackageManager::kAndroid:
-#if defined(ANDROID)
-      return std::make_shared<AndroidManager>(pconfig, bconfig, storage, http);
-#else
-      throw std::runtime_error("aktualizr was compiled without android support!");
-#endif
-    case PackageManager::kOstreeDockerApp:
-#if defined(BUILD_DOCKERAPP) && defined(BUILD_OSTREE)
-      return std::make_shared<DockerAppManager>(pconfig, bconfig, storage, http);
-#else
-      throw std::runtime_error("aktualizr was compiled without ostree+docker-app support!");
-#endif
-    case PackageManager::kNone:
-      return std::make_shared<PackageManagerFake>(pconfig, bconfig, storage, http);
-    default:
-      LOG_ERROR << "Unrecognized package manager type: " << static_cast<int>(pconfig.type);
-      return std::shared_ptr<PackageManagerInterface>();  // NULL-equivalent
+    const PackageConfig &pconfig, const BootloaderConfig &bconfig, const std::shared_ptr<INvStorage> &storage,
+    const std::shared_ptr<HttpInterface> &http) {
+  for (const auto &b : *registered_pkgms_) {
+    if (b.first == pconfig.type.c_str()) {
```

Comment:
I'm guessing there's something subtle with C++ here, but I'm not sure I see why you are call `.c_str()`?

Reply:
Haha no, it's just an artifact of a previous iteration. Good catch, thanks!